src/core/analysis/mcmc/MarginalLikelihoodEstimator.cpp

```diff
@@ -135,10 +142,379 @@ MarginalLikelihoodEstimator::MarginalLikelihoodEstimator(const path &fn, const s
+ * @return The effective sample size
+ */
+double MarginalLikelihoodEstimator::getESS(const std::vector<double> values) const
+{
+    size_t MAX_LAG = 1000;
+    size_t samples = values.size();
+    size_t maxLag = (samples - 1 < MAX_LAG ? samples - 1 : MAX_LAG);
+    
+    // get the mean
+    double mean = 0.0;
+    for (size_t i = 0; i < samples; ++i)
+    {
+        mean += values[i];
+    }
+    mean /= samples;
+    
+    double* gammaStat = new double[maxLag];
+    for (size_t j = 0; j < maxLag; j++)
+    {
+        gammaStat[j] = 0;
+    }
+    double varStat = 0.0;
+    
+    for (size_t lag = 0; lag < maxLag; lag++) {
+        for (size_t j = 0; j < samples - lag; j++) {
+            double del1 = values[j] - mean;
+            double del2 = values[j + lag] - mean;
+            gammaStat[lag] += (del1 * del2);
+        }
+
+        gammaStat[lag] /= ((double) (samples - lag));
+
+        if (lag == 0) {
+            varStat = gammaStat[0];
+        } else if (lag % 2 == 0) {
+            if (gammaStat[lag - 1] + gammaStat[lag] > 0) {
+                varStat += 2.0 * (gammaStat[lag - 1] + gammaStat[lag]);
+            } else {
+                maxLag = lag;
+            }
+        }
+    }
+    
+    // autocorrelation time
+    double act = varStat / gammaStat[0];
+
+    // effective sample size
+    double ess = samples / act;
+    
+    return ess;
+}
+
+
+std::int64_t MarginalLikelihoodEstimator::offsetModulo(std::int64_t i, std::int64_t n) const
+{
+    return 1 + (i - 1) % n;
+}
+
+
+/**
+ * Get the indices of those elements of the vector of sampled log likelihoods that will be used for a given block bootstrap replicate.
+ * The function is based on mcmc3r:::make.ends() (see https://github.com/dosreislab/mcmc3r/blob/main/R/marginal-lhd.R#L493),
+ * whose code was in turn lifted without modifications from the eponymous function in the package "boot", written by Angelo J. Canty
+ * and corrected by B. D. Ripley (see https://github.com/cran/boot/blob/master/R/bootfuns.q#L3203). It takes
+ * as its argument (1) a pair of integers representing the start index and length of a given block, and (2) the total length of the time
+ * series from which the blocks are to be drawn. The series is treated as circular, i.e., if we reach the end before the number of selected
+ * indices has reached the length of the block, we keep adding indices from the beginning,
+ *
+ * @param a Pair of integers denoting the start index of a block and its length, respectively
+ * @param n Total length of the time series
+ * @return Vector of indices to select from a time series (of sampled log likelihoods)
+ */
+std::vector<std::int64_t> MarginalLikelihoodEstimator::getIndices(std::pair<std::int64_t, std::int64_t> a, std::int64_t n) const
+{
+    std::vector<std::int64_t> out;
+    if (a.second != 0)
+    {
+        std::vector<std::int64_t> seq( a.second );
+        for (size_t i = 0; i < a.second; i++)
+        {
+            seq[i] = a.first + i;
```

Comment:
I'm not completely sure what is the point of the vector seq here

Reply:
You are correct; there was none. Removed.